hibernate-core/src/main/java/org/hibernate/type/descriptor/java/JdbcTimestampJavaType.java

```diff
@@ -130,6 +130,18 @@ public Object unwrap(Date value, Class type, WrapperOptions options) {
 					: new Timestamp( value.getTime() );
 		}
 
+		if ( java.sql.Date.class.isAssignableFrom( type ) ) {
+			return value instanceof java.sql.Date
+					? ( java.sql.Date ) value
+					: new java.sql.Date( value.getTime() );
+		}
+
+		if ( java.sql.Time.class.isAssignableFrom( type ) ) {
+			return value instanceof java.sql.Time
+					? ( java.sql.Time ) value
+					: new java.sql.Time( value.getTime() );
```

Comment:
```suggestion
					: new java.sql.Time( value.getTime() % 86_400_000 );
```

Reply:
I was planning to make separate PR for this, but I will change that on all places where it should be fixed. Or, at least all that I was able to locate.

Reply:
I can do this, but this is wrong since java.sql.Time is both time zone and date dependent. It is just subclass of java.util.Date that is hidding date related methods by throwing UnsupportedOperationException.
By using millis modulo one day as java.sql.Time constructor argument, result will be local time at system default time zone, but on January 1st 1970.

Reply:
> result will be local time at system default time zone, but on January 1st 1970.

Which is what the `java.sql.Time` [javadoc](https://docs.oracle.com/javase/8/docs/api/java/sql/Time.html) says it should always be. Also, we were also already applying the modulo before your change, so I believe it's the correct choice to do if the requested wrap type is explicitly `Time`.

Reply:
Are you sure? Shouldn't be ȷava.sql.Time` equivalent to `java.time.LocalTime` ?

Reply:
Anyway, I've changed all that and still not sure that this logic is correct, but will leave that for some other issue.

Reply:
@mbladel Problem with current conversion of (epoch) time in milliseconds to `ȷava.sql.Time` is that offset at any date/time may not be equal to one on January 1st, 1970. I numerous time zones there are usually two transitions ("summer time" and "winter time"), so at least one of those period have offset different from one in 1970.

I've added test class `org.hibernate.orm.test.mapping.type.java.GoodBadUglyTest` (disabled by default to not make build fail) which is checking two simple assumptions:
- that saved time must be equal to one loaded from DB
- that time at midnight should always be 00:00:00

Time used for testing is summer solstice of this year (which will make sure that DST is on) and three different zones:
- "good" - one where offset is equal to one at 1970; this will pass for all databases
- "bad" - one where offset is not equal to one at 1970; this will pass for some databases, but fail for others
- "ugly" - where there was "gap transition" at midnight of January 1st 1970; here tests for any time between midnight and 1 AM will fail; this time zones will disappear once when time zone database in JDK is updated to 2024b, but this will hardly be before JDK 24 is out